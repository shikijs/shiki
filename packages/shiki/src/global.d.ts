declare global {
  var __BROWSER__: boolean

  // These will not be defined for Node which don't have "DOM" in their lib.
  // So declare the minimal interface we need.
  // They are defined here and not in `loader.ts` to avoid getting injected in the built code.
  interface Window {
    WorkerGlobalScope: any
  }

  var self: Window & typeof globalThis
  function fetch(url: string): Promise<Response>
}

// This export is here so that all variables above would be added to the global scope
export {}
