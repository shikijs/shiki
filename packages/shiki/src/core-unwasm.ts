/**
 * In environments where WebAssembly can be treated as native ESM and https://github.com/unjs/unwasm,
 * We add the wasm file as the dependency so users don't need to call `loadWasm` manually.
 */

import { setDefaultWasmLoader } from '@shikijs/engine-oniguruma'

setDefaultWasmLoader(() => import('shiki/wasm'))

export * from '@shikijs/core'
