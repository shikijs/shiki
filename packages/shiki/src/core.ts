export type { CssVariablesThemeOptions } from './theme-css-variables'

export { createCssVariablesTheme } from './theme-css-variables'
export * from '@shikijs/core'
