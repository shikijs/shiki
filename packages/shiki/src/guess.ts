import type { HighlighterGeneric } from '@shikijs/types'

export function guessEmbeddedLanguages(
  code: string,
  _lang: string | undefined,
  shiki: HighlighterGeneric<any, any>,
): string[] {
  const langs = new Set<string>()
  // For HTML code blocks like Vue SFC
  for (const match of code.matchAll(/lang=["']([\w-]+)["']/g)) {
    langs.add(match[1])
  }
  // For markdown code blocks
  for (const match of code.matchAll(/(?:```|~~~)([\w-]+)/g)) {
    langs.add(match[1])
  }
  // For latex
  for (const match of code.matchAll(/\\begin\{([\w-]+)\}/g)) {
    langs.add(match[1])
  }

  // Only include known languages
  const bundle = shiki.getBundledLanguages()
  return Array.from(langs)
    .filter(l => l && bundle[l])
}
