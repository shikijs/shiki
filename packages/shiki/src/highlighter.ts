import type {
  Highlighter,
  HighlighterOptions,
  CodeToHtmlOptions,
  ILanguageRegistration,
  IShikiTheme,
  IThemeRegistration,
  StringLiteralUnion,
  AnsiToHtmlOptions
} from './types'
import { Resolver } from './resolver'
import { tokenizeWithTheme } from './themedTokenizer'
import { tokenizeAnsiWithTheme } from './ansiThemedTokenizer'
import { renderToHtml } from './renderer'

import { getOniguruma, WASM_PATH } from './loader'
import { Lang, languages as BUNDLED_LANGUAGES } from './languages'
import { Registry } from './registry'
import { Theme } from './themes'
import { namedColors as ansiNamedColors } from 'ansi-sequence-parser'

function resolveLang(lang: ILanguageRegistration | Lang) {
  return typeof lang === 'string'
    ? BUNDLED_LANGUAGES.find(l => l.id === lang || l.aliases?.includes(lang))
    : lang
}

function resolveOptions(options: HighlighterOptions) {
  let _languages: ILanguageRegistration[] = BUNDLED_LANGUAGES
  let _themes: IThemeRegistration[] = options.themes || []
  let _wasmPath: string = options.paths?.wasm
    ? options.paths.wasm.endsWith('/')
      ? options.paths.wasm
      : options.paths.wasm + '/'
    : WASM_PATH

  if (options.langs) {
    _languages = options.langs.map(resolveLang)
  }
  if (options.theme) {
    _themes.unshift(options.theme)
  }
  if (!_themes.length) {
    _themes = ['nord']
  }

  return { _languages, _themes, _wasmPath }
}

function generateDefaultColorReplacements() {
  const replacements: Record<string, string> = {
    '#000001': 'var(--shiki-color-text)',
    '#000002': 'var(--shiki-color-background)',
    '#000004': 'var(--shiki-token-constant)',
    '#000005': 'var(--shiki-token-string)',
    '#000006': 'var(--shiki-token-comment)',
    '#000007': 'var(--shiki-token-keyword)',
    '#000008': 'var(--shiki-token-parameter)',
    '#000009': 'var(--shiki-token-function)',
    '#000010': 'var(--shiki-token-string-expression)',
    '#000011': 'var(--shiki-token-punctuation)',
    '#000012': 'var(--shiki-token-link)'
  }

  // Generates a color replacement for each ANSI color with the form #A00000, #A00001, etc.
  // Note that dimmed colors will use variables with the name var(--shiki-color-ansi-<color>-dim)
  for (let i = 0; i < ansiNamedColors.length; i++) {
    const code = `#A${i.toString().padStart(5, '0')}`
    const colorNameKebab = ansiNamedColors[i].replace(/([a-z])([A-Z])/g, '$1-$2').toLowerCase()
    replacements[code] = `var(--shiki-color-ansi-${colorNameKebab})`
  }

  return replacements
}

export async function getHighlighter(options: HighlighterOptions): Promise<Highlighter> {
  const { _languages, _themes, _wasmPath } = resolveOptions(options)
  const _resolver = new Resolver(getOniguruma(_wasmPath), 'vscode-oniguruma')
  const _registry = new Registry(_resolver)

  if (options.paths?.themes) {
    _registry.themesPath = options.paths.themes.endsWith('/')
      ? options.paths.themes
      : options.paths.themes + '/'
  }

  if (options.paths?.languages) {
    _resolver.languagesPath = options.paths.languages.endsWith('/')
      ? options.paths.languages
      : options.paths.languages + '/'
  }

  const themes = await _registry.loadThemes(_themes)
  const _defaultTheme = themes[0]
  let _currentTheme: IShikiTheme | undefined
  await _registry.loadLanguages(_languages)

  /**
   * Shiki was designed for VS Code, so CSS variables are not currently supported.
   * See: https://github.com/shikijs/shiki/pull/212#issuecomment-906924986
   *
   * Instead, we work around this by using valid hex color codes as lookups in a
   * final "repair" step which translates those codes to the correct CSS variables.
   */
  let COLOR_REPLACEMENTS = generateDefaultColorReplacements()
  function setColorReplacements(map: Record<string, string>) {
    COLOR_REPLACEMENTS = map
  }
  function fixCssVariablesTheme(theme: IShikiTheme, colorMap: string[]) {
    theme.bg = COLOR_REPLACEMENTS[theme.bg] || theme.bg
    theme.fg = COLOR_REPLACEMENTS[theme.fg] || theme.fg

    Object.entries(theme.colors).forEach(([key, value]) => {
      theme.colors[key] = COLOR_REPLACEMENTS[value] || value
    })

    colorMap.forEach((val, i) => {
      colorMap[i] = COLOR_REPLACEMENTS[val] || val
    })
  }

  function getTheme(theme?: IThemeRegistration) {
    const _theme = theme ? _registry.getTheme(theme) : _defaultTheme
    if (!_theme) {
      throw Error(`No theme registration for ${theme}`)
    }
    if (!_currentTheme || _currentTheme.name !== _theme.name) {
      _registry.setTheme(_theme)
      _currentTheme = _theme
    }
    const _colorMap = _registry.getColorMap()
    if (_theme.type === 'css') {
      fixCssVariablesTheme(_theme, _colorMap)
    }
    return { _theme, _colorMap }
  }

  function getGrammar(lang: string) {
    const _grammar = _registry.getGrammar(lang)
    if (!_grammar) {
      throw Error(`No language registration for ${lang}`)
    }
    return { _grammar }
  }

  function codeToThemedTokens(
    code: string,
    lang = 'text',
    theme?: IThemeRegistration,
    options = { includeExplanation: true }
  ) {
    if (isPlaintext(lang)) {
      const lines = code.split(/\r\n|\r|\n/)
      return [...lines.map(line => [{ content: line }])]
    }
    const { _grammar } = getGrammar(lang)
    const { _theme, _colorMap } = getTheme(theme)
    return tokenizeWithTheme(_theme, _colorMap, code, _grammar, options)
  }

  function ansiToThemedTokens(ansi: string, theme?: IThemeRegistration) {
    const { _theme } = getTheme(theme)
    return tokenizeAnsiWithTheme(_theme, ansi)
  }

  function codeToHtml(code: string, options?: CodeToHtmlOptions): string
  function codeToHtml(
    code: string,
    lang?: StringLiteralUnion<Lang>,
    theme?: StringLiteralUnion<Theme>
  ): string
  function codeToHtml(
    code: string,
    arg1: StringLiteralUnion<Lang> | CodeToHtmlOptions = 'text',
    arg2?: StringLiteralUnion<Theme>
  ) {
    let options: CodeToHtmlOptions

    // codeToHtml(code, options?) overload
    if (typeof arg1 === 'object') {
      options = arg1
    }
    // codeToHtml(code, lang?, theme?) overload
    else {
      options = {
        lang: arg1,
        theme: arg2
      }
    }

    const tokens = codeToThemedTokens(code, options.lang, options.theme, {
      includeExplanation: false
    })
    const { _theme } = getTheme(options.theme)
    return renderToHtml(tokens, {
      fg: _theme.fg,
      bg: _theme.bg,
      lineOptions: options?.lineOptions,
      themeName: _theme.name
    })
  }

  function ansiToHtml(ansi: string, options?: AnsiToHtmlOptions) {
    const tokens = ansiToThemedTokens(ansi, options?.theme)
    const { _theme } = getTheme(options?.theme)
    return renderToHtml(tokens, {
      fg: _theme.fg,
      bg: _theme.bg,
      lineOptions: options?.lineOptions,
      themeName: _theme.name
    })
  }

  async function loadTheme(theme: IShikiTheme | Theme) {
    await _registry.loadTheme(theme)
  }

  async function loadLanguage(lang: ILanguageRegistration | Lang) {
    const _lang = resolveLang(lang)
    _resolver.addLanguage(_lang)
    await _registry.loadLanguage(_lang)
  }

  function getLoadedThemes() {
    return _registry.getLoadedThemes()
  }

  function getLoadedLanguages() {
    return _registry.getLoadedLanguages()
  }

  function getBackgroundColor(theme?: IThemeRegistration) {
    const { _theme } = getTheme(theme)
    return _theme.bg
  }

  function getForegroundColor(theme?: IThemeRegistration) {
    const { _theme } = getTheme(theme)
    return _theme.fg
  }

  return {
    codeToThemedTokens,
    codeToHtml,
    ansiToThemedTokens,
    ansiToHtml,
    getTheme: (theme: IThemeRegistration) => {
      return getTheme(theme)._theme
    },
    loadTheme,
    loadLanguage,
    getBackgroundColor,
    getForegroundColor,
    getLoadedThemes,
    getLoadedLanguages,
    setColorReplacements
  }
}

function isPlaintext(lang: string | null | undefined) {
  return !lang || ['plaintext', 'txt', 'text'].includes(lang)
}
