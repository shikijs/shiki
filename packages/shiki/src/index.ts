export * from './bundle-full'

export type { BuiltinLanguage, BuiltinTheme } from './types'
export { createJavaScriptRegexEngine, defaultJavaScriptRegexConstructor } from '@shikijs/engine-javascript'
export { createOnigurumaEngine, loadWasm } from '@shikijs/engine-oniguruma'
