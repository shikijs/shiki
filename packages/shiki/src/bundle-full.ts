import type { CreateHighlighterFactory, HighlighterGeneric } from '@shikijs/types'
import type {} from 'hast'
import { bundledLanguages } from './assets/langs-bundle-full'
import { createdBundledHighlighter, createSingletonShorthands, createWasmOnigEngine } from './core'
import { bundledThemes } from './themes'
import { getWasmInlined } from './wasm-dynamic'
import type { BundledLanguage } from './assets/langs-bundle-full'
import type { BundledTheme } from './themes'

export * from './assets/langs-bundle-full'
export * from './core'
export * from './themes'

export { getWasmInlined }

export type Highlighter = HighlighterGeneric<BundledLanguage, BundledTheme>

/**
 * Initiate a highlighter instance and load the specified languages and themes.
 * Later it can be used synchronously to highlight code.
 *
 * Importing this function will bundle all languages and themes.
 * @see https://shiki.style/guide/bundles#shiki-bundle-full
 *
 * For granular control over the bundle, check:
 * @see https://shiki.style/guide/bundles#fine-grained-bundle
 */
export const createHighlighter = /* @__PURE__ */ createdBundledHighlighter<
  BundledLanguage,
  BundledTheme
>({
  langs: bundledLanguages,
  themes: bundledThemes,
  engine: () => createWasmOnigEngine(getWasmInlined),
})

export const {
  codeToHtml,
  codeToHast,
  codeToTokens,
  codeToTokensBase,
  codeToTokensWithThemes,
  getSingletonHighlighter,
  getLastGrammarState,
} = /* @__PURE__ */ createSingletonShorthands<
  BundledLanguage,
  BundledTheme
>(
  createHighlighter,
)

/**
 * @deprecated Use `createHighlighter` or `getSingletonHighlighter` instead.
 */
export const getHighlighter: CreateHighlighterFactory<BundledLanguage, BundledTheme> = /* @__PURE__ */ (options) => {
  // TODO: next:  console.warn('`getHighlighter` is deprecated. Use `createHighlighter` or `getSingletonHighlighter` instead.')
  return createHighlighter(options)
}
