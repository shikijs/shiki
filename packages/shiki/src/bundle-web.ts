import type { CreateHighlighterFactory, HighlighterGeneric } from '@shikijs/types'
import type {} from 'hast'
import type { BundledLanguage } from './langs-bundle-web'
import type { BundledTheme } from './themes'
import { createdBundledHighlighter, createSingletonShorthands, warnDeprecated } from './core'
import { createOnigurumaEngine } from './engine-oniguruma'
import { bundledLanguages } from './langs-bundle-web'
import { bundledThemes } from './themes'

export * from './core'
export * from './langs-bundle-web'
export * from './themes'
export { getWasmInlined } from './wasm-dynamic'

export type Highlighter = HighlighterGeneric<BundledLanguage, BundledTheme>

/**
 * Initiate a highlighter instance and load the specified languages and themes.
 * Later it can be used synchronously to highlight code.
 *
 * Importing this function will bundle all languages and themes.
 * @see https://shiki.style/guide/bundles#shiki-bundle-web
 *
 * For granular control over the bundle, check:
 * @see https://shiki.style/guide/bundles#fine-grained-bundle
 */
export const createHighlighter = /* @__PURE__ */ createdBundledHighlighter<
  BundledLanguage,
  BundledTheme
>({
  langs: bundledLanguages,
  themes: bundledThemes,
  engine: () => createOnigurumaEngine(import('shiki/wasm')),
})

export const {
  codeToHtml,
  codeToHast,
  codeToTokensBase,
  codeToTokens,
  codeToTokensWithThemes,
  getSingletonHighlighter,
  getLastGrammarState,
} = /* @__PURE__ */ createSingletonShorthands<
  BundledLanguage,
  BundledTheme
>(
  createHighlighter,
)

/**
 * @deprecated Use `createHighlighter` or `getSingletonHighlighter` instead.
 */
export const getHighlighter: CreateHighlighterFactory<BundledLanguage, BundledTheme> = /* @__PURE__ */ (options) => {
  warnDeprecated('`getHighlighter` is deprecated. Use `createHighlighter` or `getSingletonHighlighter` instead.')
  return createHighlighter(options)
}
