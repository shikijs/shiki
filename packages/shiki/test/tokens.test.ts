import { expect, it } from 'vitest'
import { codeToTokensBase } from '../src'

it('includeExplanation', async () => {
  const code = 'console.log("hello")'
  const caseFalse = await codeToTokensBase(code, { lang: 'js', theme: 'vitesse-dark', includeExplanation: false })
  const caseTrue = await codeToTokensBase(code, { lang: 'js', theme: 'vitesse-dark', includeExplanation: true })
  const caseScopeName = await codeToTokensBase(code, { lang: 'js', theme: 'vitesse-dark', includeExplanation: 'scopeName' })

  expect(caseFalse).toMatchSnapshot('false')
  expect(caseTrue).toMatchSnapshot('true')
  expect(caseScopeName).toMatchSnapshot('scopeName')
})
