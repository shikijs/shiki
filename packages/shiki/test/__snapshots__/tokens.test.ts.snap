// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`includeExplanation > false 1`] = `
[
  [
    {
      "color": "#BD976A",
      "content": "console",
      "fontStyle": 0,
      "offset": 0,
    },
    {
      "color": "#666666",
      "content": ".",
      "fontStyle": 0,
      "offset": 7,
    },
    {
      "color": "#80A665",
      "content": "log",
      "fontStyle": 0,
      "offset": 8,
    },
    {
      "color": "#666666",
      "content": "(",
      "fontStyle": 0,
      "offset": 11,
    },
    {
      "color": "#C98A7D77",
      "content": """,
      "fontStyle": 0,
      "offset": 12,
    },
    {
      "color": "#C98A7D",
      "content": "hello",
      "fontStyle": 0,
      "offset": 13,
    },
    {
      "color": "#C98A7D77",
      "content": """,
      "fontStyle": 0,
      "offset": 18,
    },
    {
      "color": "#666666",
      "content": ")",
      "fontStyle": 0,
      "offset": 19,
    },
  ],
]
`;

exports[`includeExplanation > scopeName 1`] = `
[
  [
    {
      "color": "#BD976A",
      "content": "console",
      "explanation": [
        {
          "content": "console",
          "scopes": [
            {
              "scopeName": "source.js",
            },
            {
              "scopeName": "meta.function-call.js",
            },
            {
              "scopeName": "variable.other.object.js",
            },
          ],
        },
      ],
      "fontStyle": 0,
      "offset": 0,
    },
    {
      "color": "#666666",
      "content": ".",
      "explanation": [
        {
          "content": ".",
          "scopes": [
            {
              "scopeName": "source.js",
            },
            {
              "scopeName": "meta.function-call.js",
            },
            {
              "scopeName": "punctuation.accessor.js",
            },
          ],
        },
      ],
      "fontStyle": 0,
      "offset": 7,
    },
    {
      "color": "#80A665",
      "content": "log",
      "explanation": [
        {
          "content": "log",
          "scopes": [
            {
              "scopeName": "source.js",
            },
            {
              "scopeName": "meta.function-call.js",
            },
            {
              "scopeName": "entity.name.function.js",
            },
          ],
        },
      ],
      "fontStyle": 0,
      "offset": 8,
    },
    {
      "color": "#666666",
      "content": "(",
      "explanation": [
        {
          "content": "(",
          "scopes": [
            {
              "scopeName": "source.js",
            },
            {
              "scopeName": "meta.brace.round.js",
            },
          ],
        },
      ],
      "fontStyle": 0,
      "offset": 11,
    },
    {
      "color": "#C98A7D77",
      "content": """,
      "explanation": [
        {
          "content": """,
          "scopes": [
            {
              "scopeName": "source.js",
            },
            {
              "scopeName": "string.quoted.double.js",
            },
            {
              "scopeName": "punctuation.definition.string.begin.js",
            },
          ],
        },
      ],
      "fontStyle": 0,
      "offset": 12,
    },
    {
      "color": "#C98A7D",
      "content": "hello",
      "explanation": [
        {
          "content": "hello",
          "scopes": [
            {
              "scopeName": "source.js",
            },
            {
              "scopeName": "string.quoted.double.js",
            },
          ],
        },
      ],
      "fontStyle": 0,
      "offset": 13,
    },
    {
      "color": "#C98A7D77",
      "content": """,
      "explanation": [
        {
          "content": """,
          "scopes": [
            {
              "scopeName": "source.js",
            },
            {
              "scopeName": "string.quoted.double.js",
            },
            {
              "scopeName": "punctuation.definition.string.end.js",
            },
          ],
        },
      ],
      "fontStyle": 0,
      "offset": 18,
    },
    {
      "color": "#666666",
      "content": ")",
      "explanation": [
        {
          "content": ")",
          "scopes": [
            {
              "scopeName": "source.js",
            },
            {
              "scopeName": "meta.brace.round.js",
            },
          ],
        },
      ],
      "fontStyle": 0,
      "offset": 19,
    },
  ],
]
`;

exports[`includeExplanation > true 1`] = `
[
  [
    {
      "color": "#BD976A",
      "content": "console",
      "explanation": [
        {
          "content": "console",
          "scopes": [
            {
              "scopeName": "source.js",
              "themeMatches": [],
            },
            {
              "scopeName": "meta.function-call.js",
              "themeMatches": [],
            },
            {
              "scopeName": "variable.other.object.js",
              "themeMatches": [
                {
                  "scope": [
                    "variable",
                    "identifier",
                  ],
                  "settings": {
                    "foreground": "#bd976a",
                  },
                },
              ],
            },
          ],
        },
      ],
      "fontStyle": 0,
      "offset": 0,
    },
    {
      "color": "#666666",
      "content": ".",
      "explanation": [
        {
          "content": ".",
          "scopes": [
            {
              "scopeName": "source.js",
              "themeMatches": [],
            },
            {
              "scopeName": "meta.function-call.js",
              "themeMatches": [],
            },
            {
              "scopeName": "punctuation.accessor.js",
              "themeMatches": [
                {
                  "scope": [
                    "delimiter.bracket",
                    "delimiter",
                    "invalid.illegal.character-not-allowed-here.html",
                    "keyword.operator.rest",
                    "keyword.operator.spread",
                    "keyword.operator.type.annotation",
                    "keyword.operator.relational",
                    "keyword.operator.assignment",
                    "keyword.operator.type",
                    "meta.brace",
                    "meta.tag.block.any.html",
                    "meta.tag.inline.any.html",
                    "meta.tag.structure.input.void.html",
                    "meta.type.annotation",
                    "meta.embedded.block.github-actions-expression",
                    "storage.type.function.arrow",
                    "meta.objectliteral.ts",
                    "punctuation",
                    "punctuation.definition.string.begin.html.vue",
                    "punctuation.definition.string.end.html.vue",
                  ],
                  "settings": {
                    "foreground": "#666666",
                  },
                },
              ],
            },
          ],
        },
      ],
      "fontStyle": 0,
      "offset": 7,
    },
    {
      "color": "#80A665",
      "content": "log",
      "explanation": [
        {
          "content": "log",
          "scopes": [
            {
              "scopeName": "source.js",
              "themeMatches": [],
            },
            {
              "scopeName": "meta.function-call.js",
              "themeMatches": [],
            },
            {
              "scopeName": "entity.name.function.js",
              "themeMatches": [
                {
                  "scope": [
                    "entity",
                    "entity.name",
                  ],
                  "settings": {
                    "foreground": "#80a665",
                  },
                },
                {
                  "scope": "entity.name.function",
                  "settings": {
                    "foreground": "#80a665",
                  },
                },
              ],
            },
          ],
        },
      ],
      "fontStyle": 0,
      "offset": 8,
    },
    {
      "color": "#666666",
      "content": "(",
      "explanation": [
        {
          "content": "(",
          "scopes": [
            {
              "scopeName": "source.js",
              "themeMatches": [],
            },
            {
              "scopeName": "meta.brace.round.js",
              "themeMatches": [
                {
                  "scope": [
                    "delimiter.bracket",
                    "delimiter",
                    "invalid.illegal.character-not-allowed-here.html",
                    "keyword.operator.rest",
                    "keyword.operator.spread",
                    "keyword.operator.type.annotation",
                    "keyword.operator.relational",
                    "keyword.operator.assignment",
                    "keyword.operator.type",
                    "meta.brace",
                    "meta.tag.block.any.html",
                    "meta.tag.inline.any.html",
                    "meta.tag.structure.input.void.html",
                    "meta.type.annotation",
                    "meta.embedded.block.github-actions-expression",
                    "storage.type.function.arrow",
                    "meta.objectliteral.ts",
                    "punctuation",
                    "punctuation.definition.string.begin.html.vue",
                    "punctuation.definition.string.end.html.vue",
                  ],
                  "settings": {
                    "foreground": "#666666",
                  },
                },
              ],
            },
          ],
        },
      ],
      "fontStyle": 0,
      "offset": 11,
    },
    {
      "color": "#C98A7D77",
      "content": """,
      "explanation": [
        {
          "content": """,
          "scopes": [
            {
              "scopeName": "source.js",
              "themeMatches": [],
            },
            {
              "scopeName": "string.quoted.double.js",
              "themeMatches": [
                {
                  "scope": [
                    "string",
                    "string punctuation.section.embedded source",
                    "attribute.value",
                  ],
                  "settings": {
                    "foreground": "#c98a7d",
                  },
                },
              ],
            },
            {
              "scopeName": "punctuation.definition.string.begin.js",
              "themeMatches": [
                {
                  "scope": [
                    "delimiter.bracket",
                    "delimiter",
                    "invalid.illegal.character-not-allowed-here.html",
                    "keyword.operator.rest",
                    "keyword.operator.spread",
                    "keyword.operator.type.annotation",
                    "keyword.operator.relational",
                    "keyword.operator.assignment",
                    "keyword.operator.type",
                    "meta.brace",
                    "meta.tag.block.any.html",
                    "meta.tag.inline.any.html",
                    "meta.tag.structure.input.void.html",
                    "meta.type.annotation",
                    "meta.embedded.block.github-actions-expression",
                    "storage.type.function.arrow",
                    "meta.objectliteral.ts",
                    "punctuation",
                    "punctuation.definition.string.begin.html.vue",
                    "punctuation.definition.string.end.html.vue",
                  ],
                  "settings": {
                    "foreground": "#666666",
                  },
                },
                {
                  "scope": [
                    "punctuation.definition.string",
                  ],
                  "settings": {
                    "foreground": "#c98a7d77",
                  },
                },
              ],
            },
          ],
        },
      ],
      "fontStyle": 0,
      "offset": 12,
    },
    {
      "color": "#C98A7D",
      "content": "hello",
      "explanation": [
        {
          "content": "hello",
          "scopes": [
            {
              "scopeName": "source.js",
              "themeMatches": [],
            },
            {
              "scopeName": "string.quoted.double.js",
              "themeMatches": [
                {
                  "scope": [
                    "string",
                    "string punctuation.section.embedded source",
                    "attribute.value",
                  ],
                  "settings": {
                    "foreground": "#c98a7d",
                  },
                },
              ],
            },
          ],
        },
      ],
      "fontStyle": 0,
      "offset": 13,
    },
    {
      "color": "#C98A7D77",
      "content": """,
      "explanation": [
        {
          "content": """,
          "scopes": [
            {
              "scopeName": "source.js",
              "themeMatches": [],
            },
            {
              "scopeName": "string.quoted.double.js",
              "themeMatches": [
                {
                  "scope": [
                    "string",
                    "string punctuation.section.embedded source",
                    "attribute.value",
                  ],
                  "settings": {
                    "foreground": "#c98a7d",
                  },
                },
              ],
            },
            {
              "scopeName": "punctuation.definition.string.end.js",
              "themeMatches": [
                {
                  "scope": [
                    "delimiter.bracket",
                    "delimiter",
                    "invalid.illegal.character-not-allowed-here.html",
                    "keyword.operator.rest",
                    "keyword.operator.spread",
                    "keyword.operator.type.annotation",
                    "keyword.operator.relational",
                    "keyword.operator.assignment",
                    "keyword.operator.type",
                    "meta.brace",
                    "meta.tag.block.any.html",
                    "meta.tag.inline.any.html",
                    "meta.tag.structure.input.void.html",
                    "meta.type.annotation",
                    "meta.embedded.block.github-actions-expression",
                    "storage.type.function.arrow",
                    "meta.objectliteral.ts",
                    "punctuation",
                    "punctuation.definition.string.begin.html.vue",
                    "punctuation.definition.string.end.html.vue",
                  ],
                  "settings": {
                    "foreground": "#666666",
                  },
                },
                {
                  "scope": [
                    "punctuation.definition.string",
                  ],
                  "settings": {
                    "foreground": "#c98a7d77",
                  },
                },
              ],
            },
          ],
        },
      ],
      "fontStyle": 0,
      "offset": 18,
    },
    {
      "color": "#666666",
      "content": ")",
      "explanation": [
        {
          "content": ")",
          "scopes": [
            {
              "scopeName": "source.js",
              "themeMatches": [],
            },
            {
              "scopeName": "meta.brace.round.js",
              "themeMatches": [
                {
                  "scope": [
                    "delimiter.bracket",
                    "delimiter",
                    "invalid.illegal.character-not-allowed-here.html",
                    "keyword.operator.rest",
                    "keyword.operator.spread",
                    "keyword.operator.type.annotation",
                    "keyword.operator.relational",
                    "keyword.operator.assignment",
                    "keyword.operator.type",
                    "meta.brace",
                    "meta.tag.block.any.html",
                    "meta.tag.inline.any.html",
                    "meta.tag.structure.input.void.html",
                    "meta.type.annotation",
                    "meta.embedded.block.github-actions-expression",
                    "storage.type.function.arrow",
                    "meta.objectliteral.ts",
                    "punctuation",
                    "punctuation.definition.string.begin.html.vue",
                    "punctuation.definition.string.end.html.vue",
                  ],
                  "settings": {
                    "foreground": "#666666",
                  },
                },
              ],
            },
          ],
        },
      ],
      "fontStyle": 0,
      "offset": 19,
    },
  ],
]
`;
